File: scoreserver/ScoreServer.js
class ScoreServer {
    constructor() {
        this.games = [];
        window.scoreServer = this;
    }

    findGame(name) {
        return this.games.find(g => g.name.toLowerCase() === name.toLowerCase());
    }

    isHi(game, score) {
        const g = this.findGame(game);
        if (!g) return 0;
        return g.isHiScore(score) ? 1 : 0;
    }

    getTable(game) {
        let g = this.findGame(game);
        if (!g) {
            g = new GameTable(game);
            this.games.push(g);
        }
        return g.getScores();
    }

    addScore(game, name, score) {
        let g = this.findGame(game);
        if (!g) {
            g = new GameTable(game);
            this.games.push(g);
        }
        g.addScore(name, score);
    }

    handle(commandString) {
        const parts = commandString.split('#');
        const command = parts[0].toLowerCase();
        const game = parts[1];
        switch (command) {
            case 'get':
                return this.getTable(game);
            case 'clr':
                const g = this.findGame(game);
                if (g) {
                    this.games = this.games.filter(x => x !== g);
                    this.games.push(new GameTable(game));
                }
                return '';
            case 'chk':
                if (parts.length >= 3) {
                    const score = parseInt(parts[2], 10);
                    return String(this.isHi(game, score));
                }
                return '0';
            case 'add':
                if (parts.length >= 4) {
                    const score = parseInt(parts[2], 10);
                    const name = parts[3];
                    this.addScore(game, name, score);
                }
                return '';
            default:
                return '';
        }
    }
}

class Player {
    constructor(name, score) {
        this.playerName = name;
        this.playerScore = score;
    }
}

class GameTable {
    constructor(name) {
        this.name = name;
        this.playerScores = [];
        for (let i = 0; i < 10; i++) {
            this.playerScores.push(new Player(`${name} ${i + 1}`, 0));
        }
    }

    isHiScore(score) {
        const p = this.playerScores[9];
        return score > p.playerScore;
    }

    addScore(name, score) {
        const p = new Player(name, score);
        let i = 0;
        while (i < this.playerScores.length && p.playerScore <= this.playerScores[i].playerScore) {
            i++;
        }
        this.playerScores.splice(i, 0, p);
        if (this.playerScores.length > 10) {
            this.playerScores.pop();
        }
    }

    getScores() {
        return this.playerScores
            .slice(0, 10)
            .map(p => `${p.playerName}#${p.playerScore}`)
            .join(';');
    }
}

export { ScoreServer };

File: scoreserver/ScoreServer.java
import netscape.security.PrivilegeManager;
import java.net.*;
import java.io.*;
import java.util.Vector;
import java.util.StringTokenizer;

public class ScoreServer
{
	ServerSocket serverSocket;
	Vector Games;

	class Player
	{
		String playerName;
		int playerScore;
		
		public Player(String name, int score)
		{
			playerName = name;
			playerScore = score;
		}
	}

	class GameTable
	{
		String name;
		Vector playerScores;

		public GameTable(String pName)
		{
			playerScores = new Vector();
			for(int i=0;i<10;i++)
			{
				Player p = new Player(pName + " " + (i+1), 0);
				playerScores.insertElementAt(p, i);
			}

			name = pName;
		}

		public boolean isHiScore(int score)
		{
			Player p;
			
			p = (Player)playerScores.elementAt(9);

			return(score > p.playerScore);
		}

		public void addScore(String name, int score)
		{
			Player p = new Player(name, score);
			Player hiPlayer;
			int i;

			i = 0;
			while(i < 10)
			{
				hiPlayer = (Player)playerScores.elementAt(i);
				if(p.playerScore > hiPlayer.playerScore)
				{
					playerScores.insertElementAt(p, i);
					break;
				}
				i++;
			}
		}

		public String getScores()
		{
			StringBuffer sb = new StringBuffer("");
			int i;
			Player p;

			for(i=0;i<10;i++)
			{
				p = (Player)playerScores.elementAt(i);
				if(i > 0)
				{
					sb.append(";");
				}
				sb.append(p.playerName);
				sb.append("#");
				sb.append(p.playerScore);
			}
			return sb.toString();
		}
	}

	private GameTable findGame(String game)
	{
		if(Games.size() == 0) return null;

		for(int i = 0; i < Games.size(); i++)
		{
			GameTable g = (GameTable)Games.elementAt(i);
			if(g.name.equalsIgnoreCase(game)) return g;
		}

		return null;
	}

	private int isHi(String game, int score)
	{
		GameTable g = findGame(game);
		if(g == null) return 0;
		return (g.isHiScore(score) ? 1 : 0);
	}

	private String getTable(String game)
	{
		GameTable g = findGame(game);

		if(g == null)
		{
			g = new GameTable(game);
			Games.addElement(g);
		}

		return g.getScores()+"\n";
	}

	private void addScore(String game, String pName, int pScore)
	{
		GameTable g;
		
		g = findGame(game);

		if(g == null)
		{
			g = new GameTable(game);
			Games.addElement(g);
		}

		g.addScore(pName, pScore);
	}

	public ScoreServer()
	{
		Games = new Vector();
		Socket client;
		DataInputStream dataIn;
		DataOutputStream dataOut;
		BufferedReader textIn;
		String commandString = "";
		String command = "";
		String game = "";
		String pName = "";
		String score = "";
		int pScore = 0;
		String scoreList;
		GameTable gt;

		try
		{
			serverSocket = new ServerSocket(9800);


			while(true)
			{
				client = serverSocket.accept();
				System.out.println("Client connection made");

				dataIn = new DataInputStream(client.getInputStream());
				dataOut = new DataOutputStream(client.getOutputStream());
				textIn = new BufferedReader(new InputStreamReader(client.getInputStream()));

				commandString = textIn.readLine();

				StringTokenizer st = new StringTokenizer(commandString, "#");
				
				int numTokens = st.countTokens();
				
				command = st.nextToken();
				game = st.nextToken();

				if(numTokens >= 3)
				{
					score = st.nextToken();
					pScore = Integer.parseInt(score);
				}

				if(numTokens >= 4)
				{
					pName = st.nextToken();
				}
					
				System.out.println("Command = " + command);
				if(command.equalsIgnoreCase("get"))
				{
					scoreList = getTable(game);
					dataOut.writeBytes(scoreList);
				}

				if(command.equalsIgnoreCase("clr"))
				{
					gt = findGame(game);
					if(gt != null)
					{
						Games.removeElement(gt);
						gt = new GameTable(game);
						Games.addElement(gt);
					}
				}

				if(command.equalsIgnoreCase("chk"))
				{
					dataOut.writeInt(isHi(game, pScore));
				}

				if(command.equalsIgnoreCase("add"))
				{
					addScore(game, pName, pScore);
				}
			}
		}
		catch(IOException exIO)
		{
			System.out.println("Unable to create server socket");
			System.out.println(exIO.getMessage());
		}
		
	}
	public static void main(String argv[])
	{
		ScoreServer ss = new ScoreServer();
	}
}
